spdmlib/src/common/opaque.rs

```diff
@@ -30,81 +26,156 @@ pub const SUPPORTED_VERSION_LIST_SM_DATA_ID: u8 = 0x01;
-    0x00, // padding
-];
-
-pub const RSP_DMTF_OPAQUE_DATA_VERSION_SELECTION_DSP0274_FMT1: [u8; 12] = [
-    OPAQUE_LIST_TOTAL_ELEMENTS,
-    0x00, // reserved
-    0x00, // reserved
-    0x00, // reserved
-    DMTF_ID,
-    DMTF_VENDOR_LEN,
-    0x04,
-    0x00,
-    SM_DATA_VERSION,
-    VERSION_SELECTION_SM_DATA_ID,
-    0x00,
-    DMTF_SECURE_SPDM_VERSION_11,
-];
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct GeneralOpaqueDataHeader;
+
+impl Codec for GeneralOpaqueDataHeader {
+    fn encode(&self, bytes: &mut Writer) -> Result<usize, codec::EncodeErr> {
+        let mut cnt = 0;
+
+        cnt += DMTF_SPEC_ID.encode(bytes)?;
+        cnt += DMTF_OPAQUE_VERSION.encode(bytes)?;
+        cnt += OPAQUE_LIST_TOTAL_ELEMENTS.encode(bytes)?;
+        cnt += 0u16.encode(bytes)?; // reserved
+
+        Ok(cnt)
+    }
+
+    fn read(r: &mut Reader) -> Option<Self> {
+        let spec_id = u32::read(r)?;
+        let opaque_version = u8::read(r)?;
+        let opaque_list_total_elements = u8::read(r)?;
+        u16::read(r)?; // reserved
+
+        if spec_id != DMTF_SPEC_ID
+            || opaque_version != DMTF_OPAQUE_VERSION
+            || opaque_list_total_elements != OPAQUE_LIST_TOTAL_ELEMENTS
+        {
+            None
+        } else {
+            Some(Self)
+        }
+    }
+}
+
+impl SpdmCodec for GeneralOpaqueDataHeader {
+    fn spdm_encode(&self, _context: &mut SpdmContext, bytes: &mut Writer) -> SpdmResult<usize> {
+        self.encode(bytes).map_err(|_| SPDM_STATUS_BUFFER_FULL)
+    }
+
+    fn spdm_read(_context: &mut SpdmContext, r: &mut Reader) -> Option<Self> {
+        GeneralOpaqueDataHeader::read(r)
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct FM1OpaqueDataHeader;
+
+impl Codec for FM1OpaqueDataHeader {
+    fn encode(&self, bytes: &mut Writer) -> Result<usize, codec::EncodeErr> {
+        let mut cnt = 0;
+
+        cnt += OPAQUE_LIST_TOTAL_ELEMENTS.encode(bytes)?;
+        cnt += u24::new(0).encode(bytes)?; // reserved
+
+        Ok(cnt)
+    }
+
+    fn read(r: &mut Reader) -> Option<Self> {
+        let opaque_list_total_elements = u8::read(r)?;
+        u24::read(r)?; // reserved
+
+        if opaque_list_total_elements != OPAQUE_LIST_TOTAL_ELEMENTS {
+            None
+        } else {
+            Some(Self)
+        }
+    }
+}
+
+impl SpdmCodec for FM1OpaqueDataHeader {
+    fn spdm_encode(&self, _context: &mut SpdmContext, bytes: &mut Writer) -> SpdmResult<usize> {
+        self.encode(bytes).map_err(|_| SPDM_STATUS_BUFFER_FULL)
+    }
+
+    fn spdm_read(_context: &mut SpdmContext, r: &mut Reader) -> Option<Self> {
+        FM1OpaqueDataHeader::read(r)
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum SMOpaqueDataHeader {
+    General(GeneralOpaqueDataHeader),
+    FM1(FM1OpaqueDataHeader),
+}
+
+impl Codec for SMOpaqueDataHeader {
+    fn encode(&self, bytes: &mut Writer) -> Result<usize, codec::EncodeErr> {
+        match self {
```

Comment:
I am not sure how this work.

Does the caller need to know the difference?

Reply:
> I am not sure how this work.
> 
> Does the caller need to know the difference?
> 
> I recommend to hide the info from the caller. The caller should just input SPDM version. The rest should be handled by the callee.

Suggestion applied